projects/ADuCM3029_demo_cn0397/src/cn0397_app.cpp

```diff
@@ -0,0 +1,519 @@
+NO EVENT SHALL ANALOG DEVICES, INC. OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, PUNITIVE OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, DAMAGES ARISING OUT OF CLAIMS OF INTELLECTUAL
+PROPERTY RIGHTS INFRINGEMENT; PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
+EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*****************************************************************************/
+
+#include <sys/platform.h>
+#include "cn0397_app.h"
+#include "common.h"
+#include <common/adi_error_handling.h>
+#include "base_sensor/adi_sensor_errors.h"
+
+using namespace adi_sensor_swpack;
+
+/* Peripheral advertisement mode */
+#define PERIPHERAL_ADV_MODE      ((ADI_BLE_GAP_MODE)(ADI_BLE_GAP_MODE_CONNECTABLE | ADI_BLE_GAP_MODE_DISCOVERABLE))
+
+#if(ADI_APP_USE_BLUETOOTH == 1)
+/* Static bluetooth data */
+static bool               gbConnected;
+static ADI_BLE_GAP_MODE   geMode;
+uint8_t reg_flag;
+
+ RegistrationPacket_t reg_pkt = {0x00, 3, 5, "CN0397"};
+ FieldNamePacket_t name_pkt0 = {0x01, 0, "Red Light [lux]"};
+ FieldNamePacket_t name_pkt1 = {0x01, 1, "Blue Light [lux]"};
+ FieldNamePacket_t name_pkt2 = {0x01, 2, "Green Light [lux]"};
+ DataPacket_t data_pkt = {0x02, 0, 0, 0, 0};
+
+/* Local bluetooth functions */
+static void InitBluetoothLowEnergy(void);
+static void ApplicationCallback(void * pCBParam, uint32_t Event, void * pArg);
+static void SetAdvertisingMode(void);
+static void VisibleLightBluetoothMode(VisibleLight *pLight);
+
+#else
+#ifndef ADI_DEBUG
+/* Static standalone data for release mode UART redirection*/
+static  int  BarLine[3];
+
+/* Static functions for release mode UART redirection*/
+static void CN0397_SetBar(float conc, int *line);
+#endif
+/* Local standalone functions */
+static void VisibleLightStandaloneMode(VisibleLight *pLight);
+
+#endif
+
+/* Local functions */
+static void InitSystem(void);
+static void Trap(void);
+
+/*!
+ * @brief      Main
+ *
+ * @details    Application entry point.
+ *
+ * @param [in] argc : Number of arguments (unused)
+ *
+ * @param [in] argv : Arguments (unused)
+ *
+ */
+int main(int argc, char *argv[])
+{
+    CN0397          cn0397;
+    VisibleLight   *pLight = &cn0397;
+    SENSOR_RESULT   eSensorResult;
+
+    /* Initialize the system */
+    InitSystem();
+
+    /* Open the visible light sensor */
+    eSensorResult = pLight->open();
+
+    if(eSensorResult != SENSOR_ERROR_NONE) {
+        PRINT_SENSOR_ERROR(DEBUG_MESSAGE, eSensorResult);
+        Trap();
+    }
+
+    DEBUG_MESSAGE("Please cover the sensor for the calibrations. \r\n");
+
+    /* Delay for approx 5 seconds */
+    for(volatile uint32_t i = 0; i < 12000000u; i++);
```

Comment:
Declaring variables elsewhere but the function start is not compatible with older standard C compilers like C89 and ANSI, but they have been updated to allow this as a GNU extension. Do we want to keep the pedantic standards, or can we go with this?

Is there a better way to implement a delay here?

Reply:
For now i will leave this as is. Would require an extra C file for timers to be able to implement a different delay.